src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+                                ASSERT((*it)->_instruction == CSYNC_INSTRUCTION_REMOVE);
+                                (*it)->_instruction = CSYNC_INSTRUCTION_NONE;
+                            }
+                            delete _discoveryData->_queuedDeletedDirectories.take(originalPath);
+
+                            processRename(path);
+                            recurseQueryServer = *etag == base._etag ? ParentNotChanged : NormalQuery;
+                        }
+                        processFileFinalize(item, path, item->isDirectory(), NormalQuery, recurseQueryServer);
+                        _pendingAsyncJobs--;
+                        QTimer::singleShot(0, _discoveryData, &DiscoveryPhase::scheduleMoreJobs);
+                    });
+                    job->start();
+                    return;
+                } else {
+                    postProcessLocalNew();
+                }
+            } else {
+                postProcessLocalNew();
+            }
+        } else {
+            item->_instruction = CSYNC_INSTRUCTION_SYNC;
+            if (noServerEntry) {
+                // Special case! deleted on server, modified on client, the instruction is then NEW
+                item->_instruction = CSYNC_INSTRUCTION_NEW;
+            }
+            item->_direction = SyncFileItem::Up;
+            item->_checksumHeader.clear();
+            item->_size = localEntry.size;
+            item->_modtime = localEntry.modtime;
+            item->_previousSize = dbEntry._fileSize;
+            item->_previousModtime = dbEntry._modtime;
+            if (!_dirItem || _dirItem->_direction == SyncFileItem::Down) {
+                _childModified = true;
+            }
+
+            // Checksum comparison at this stage is only enabled for .eml files,
+            // check #4754 #4755
+            bool isEmlFile = path._original.endsWith(QLatin1String(".eml"), Qt::CaseInsensitive);
+            if (isEmlFile && dbEntry._fileSize == localEntry.size && !dbEntry._checksumHeader.isEmpty()) {
+                if (computeLocalChecksum(dbEntry._checksumHeader, _discoveryData->_localDir + path._local, item)
+                        && item->_checksumHeader == dbEntry._checksumHeader) {
+                    qCInfo(lcDisco) << "NOTE: Checksums are identical, file did not actually change: " << path._local;
+                    item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
+                }
+            }
+        }
+    } else if (_queryLocal == ParentNotChanged && dbEntry.isValid()) {
+        if (noServerEntry) {
+            // Not modified locally (ParentNotChanged), bit not on the server:  Removed on the server.
+            item->_instruction = CSYNC_INSTRUCTION_REMOVE;
+            item->_direction = SyncFileItem::Down;
+        }
+    } else if (noServerEntry) {
+        // Not locally, not on the server. The entry is stale!
+        qCInfo(lcDisco) << "Stale DB entry";
```

Comment:
I think we want to wipe the local db entry in this case (and add a remove/remove test if we don't have one yet)